core/src/main/scala/org/apache/spark/rdd/CartesianRDD.scala

```diff
@@ -71,9 +72,92 @@ class CartesianRDD[T: ClassTag, U: ClassTag](
   }
 
   override def compute(split: Partition, context: TaskContext): Iterator[(T, U)] = {
+    val blockManager = SparkEnv.get.blockManager
     val currSplit = split.asInstanceOf[CartesianPartition]
-    for (x <- rdd1.iterator(currSplit.s1, context);
-         y <- rdd2.iterator(currSplit.s2, context)) yield (x, y)
+    val blockId2 = RDDBlockId(rdd2.id, currSplit.s2.index)
+    var cachedInLocal = false
+    var holdReadLock = false
+
+    // Try to get data from the local, otherwise it will be cached to the local.
+    def getOrElseCache(
+        rdd: RDD[U],
+        partition: Partition,
+        context: TaskContext,
+        level: StorageLevel): Iterator[U] = {
+      getLocalValues() match {
+        case Some(result) =>
+          return result
+        case None => if (holdReadLock) {
+          throw new SparkException(s"get() failed for block $blockId2 even though we held a lock")
+        }
+      }
+
+      val iterator = rdd.iterator(partition, context)
+      if (rdd.getStorageLevel != StorageLevel.NONE || rdd.isCheckpointedAndMaterialized) {
+        // If the block is cached in local, wo shouldn't cache it again.
```

Comment:
@viirya Thanks for review. If the storeage level of rdd2 is not `StorageLevel.NONE`, it will cached by in the method `RDD.getOrCompute`. So I think we should cache it again, because the `blockManger.getOrElseUpdate` call the same method as `blockManager.putIterator`.